### ISSUE TITLE
PDF only generating with first table

### ISSUE THREAD

**Issue by c1am** [dompdf/dompdf#2943]
I'm running into an issue where the PDF only generates with the first table, even though the HTML object passed into it has 2 or 3 tables. I'm not sure what's happening even with a debugger.

**Comment by bsweeney** [comment]
That is _quite_ the HTML document. I'll need a bit to review what's going on there. The nested HTML elements are not helping. Removing just those extra HTML tags seems to at least address the missing content. I'm not sure there's a lot we can do for this considering the wonky (and syntactically invalid) document structure. Ideally the content of the nested HTML documents would be insert into the main document instead of the entire thing. Or using iframes to contain that content (except dompdf doesn't yet support iframes).

**Comment by c1am** [comment]
It looks a bit wonky because they're emails. I'm just not really sure what is causing it to break after the first email. There are 3 emails there and I can switch the order, but only the first one shows in the PDF render even though the HTML we're passing looks good when printed on the page or when we grab the value from the debugger.

**Comment by c1am** [comment]
Hey, just following up to see if you were able to figure out what's going on.

**Comment by c1am** [comment]
So I actually think I found the offending text. It's a set of invalid HTML tags i.e. <abc123>abc123<abc123>. Do you know of anything that works with removing invalid HTML tags/false DOM elements?

**Comment by bsweeney** [comment]
I haven't had a chance to look into the issue more.

As for your question, the easiest way to remove those tags would be to use PHP's `strip_tags` function ... except you have to supply a lost of tags that can stay. Which if we're talking about all of HTML that's a lot to specify. It's too bad you can specify which tags to strip instead.

You could just try using preg_replace:

```
$html = preg_replace('/<abc123[^>]*>(.*?)<\/abc123>/', '$1', $html);
```

Color me surprised. I would have expected Dompdf to still render that content (since 2.0.0). But that only appears to be the case if the element is contained within a `body` element. So you can get around the issue by wrapping the entire ugly multi-document things inside yet another HTML document:

```
$html = "<html><body>" . $html . "</body></html>";
``

producting something like the following ... which does render
```
<html>
<body>
<html>
<body>
<abc123>abc123<abc123>
</body>
</html>
<abc123>abc123<abc123>
</body>
</html>
```

I did that with the sample you posted earlier and it appears to render OK.

**Comment by bsweeney** [comment]
Did you try wrapping the content in `<html><body>...</html></body>` with 2.0? When I did that the full document rendered.

**Comment by bsweeney** [comment]
welp ... seems to render fine on the [debugger](https://eclecticgeek.com/dompdf/debug.php)? Maybe it's an issue with one of the underlying libraries? What version of PHP, libxml, etc?

**Comment by c1am** [comment]
Honestly, I'm not really sure what happened, but with some extensive testing and re-testing with our code, I think we were able to get it to work by wrapping the resulting HTML with the `<html><body>...</body></html>` and changing some other stuff up.

For PDFs that would be... let's say hundreds of pages long, is there a way to render the HTML in parts and then combine the PDFs together?

**Comment by bsweeney** [comment]
Yeah, long HTML documents are fairly slow to render. Particularly if they have any type of complexity (lots of elements, for example).

I've used split rendering in the past and it can significantly improve performance. The documents I rendered were fairly easily paged, so I built up pages individually and rendered them separately. If you have one large document with an easily identifiable boundary you should just PHP split on that boundary.

I used pdftk to combine the documents after rendering was complete. There are a few options built on top of FPDF/FPDI, though it's been a while since I looked at any of them. Most seem to have been abandoned. I think it would be useful to be able to render multiple HTML documents into the same PDF document via Dompdf, though I haven't really had a chance to think it through.
